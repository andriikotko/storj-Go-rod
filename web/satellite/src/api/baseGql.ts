// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import gql from 'graphql-tag';

import { apollo } from '@/utils/apollo';

/**
 * BaseGql is a graphql utility which allows to perform queries and mutations.
 */
export class BaseGql {
    /**
     * performs qraphql query.
     *
     * @param query - qraphql query
     * @param variables - variables to bind in query. null by default.
     * @throws Error
     */
    protected async query(query: string, variables: any = null): Promise<any> {
        const response: any = await apollo.query(
            {
                query: gql(query),
                variables,
                fetchPolicy: 'no-cache',
                errorPolicy: 'all',
            },
        );

        if (response.errors) {
            throw new Error(this.combineErrors(response.errors));
        }

        return response;
    }

    /**
     * performs qraphql mutation.
     *
     * @param query - qraphql query
     * @param variables - variables to bind in query. null by default.
     * @throws Error
     */
    protected async mutate(query: string, variables: any = null): Promise<any> {
        const response: any = await apollo.mutate(
            {
                mutation: gql(query),
                variables,
                fetchPolicy: 'no-cache',
                errorPolicy: 'all',
            },
        );

        if (response.errors) {
            throw new Error(this.combineErrors(response.errors));
        }

        return response;
    }

    private combineErrors(gqlError: any): string {
        return gqlError.map(err => err).join('\n');
    }
}
