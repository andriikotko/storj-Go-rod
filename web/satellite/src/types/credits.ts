// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

/**
 * Exposes all credits-related functionality.
 */
export interface CreditsApi {
    get(): Promise<CreditUsage>;
}

/**
 * CreditUsage stores information about users credits.
 * Tardigrade related logic
 */
export class CreditUsage {
    public referred: number;
    public usedCredits: number;
    public availableCredits: number;

    constructor(referred: number = 0, usedCredits: number = 0, availableCredits: number = 0) {
        this.referred = referred;
        this.usedCredits = usedCredits;
        this.availableCredits = availableCredits;
    }
}
