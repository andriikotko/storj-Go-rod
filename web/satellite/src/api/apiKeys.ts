// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { BaseGql } from '@/api/baseGql';
import { ApiKey, ApiKeyCursor, ApiKeysApi, ApiKeysPage } from '@/types/apiKeys';

/**
 * ApiKeysApiGql is a graphql implementation of ApiKeys API.
 * Exposes all apiKey-related functionality
 */
export class ApiKeysApiGql extends BaseGql implements ApiKeysApi {
    /**
     * Fetch apiKeys.
     *
     * @returns ApiKey
     * @throws Error
     */
    public async get(projectId: string, cursor: ApiKeyCursor): Promise<ApiKeysPage> {
        const query =
            `query($projectId: String!, $limit: Int!, $search: String!, $page: Int!, $order: Int!, $orderDirection: Int!) {
                project (
                    id: $projectId,
                ) {
                    apiKeys (
                        cursor: {
                            limit: $limit,
                            search: $search,
                            page: $page,
                            order: $order,
                            orderDirection: $orderDirection
                        }
                    ) {
                        apiKeys {
                            id,
                            name,
                            createdAt
                        }
                        search,
                        limit,
                        order,
                        pageCount,
                        currentPage,
                        totalCount
                    }
                }
            }`;

        const variables = {
            projectId: projectId,
            limit: cursor.limit,
            search: cursor.search,
            page: cursor.page,
            order: cursor.order,
            orderDirection: cursor.orderDirection,
        };

        const response = await this.query(query, variables);

        return this.getApiKeysPage(response.data.project.apiKeys);
    }

    /**
     * Used to create apiKey.
     *
     * @param projectId - stores current project id
     * @param name - name of apiKey that will be created
     * @returns ApiKey
     * @throws Error
     */
    public async create(projectId: string, name: string): Promise<ApiKey> {
        const query =
            `mutation($projectId: String!, $name: String!) {
                createAPIKey(
                    projectID: $projectId,
                    name: $name
                ) {
                    key,
                    keyInfo {
                        id,
                        name,
                        createdAt
                    }
                }
            }`;

        const variables = {
            projectId,
            name,
        };

        const response = await this.mutate(query, variables);
        const key: any = response.data.createAPIKey.keyInfo;
        const secret: string = response.data.createAPIKey.key;

        return new ApiKey(key.id, key.name, key.createdAt, secret);
    }

    /**
     * Used to delete apiKey.
     *
     * @param ids - ids of apiKeys that will be deleted
     * @throws Error
     */
    public async delete(ids: string[]): Promise<void> {
        const query =
            `mutation($id: [String!]!) {
                deleteAPIKeys(id: $id) {
                    id
                }
            }`;

        const variables = {
            id: ids,
        };

        const response = await this.mutate(query, variables);

        return response.data.deleteAPIKeys;
    }

    /**
     * Method for mapping api keys page from json to ApiKeysPage type.
     *
     * @param page anonymous object from json
     */
    private getApiKeysPage(page: any): ApiKeysPage {
        if (!page) {
            return new ApiKeysPage();
        }

        const apiKeysPage: ApiKeysPage = new ApiKeysPage();

        apiKeysPage.apiKeys = page.apiKeys.map(key => new ApiKey(
            key.id,
            key.name,
            new Date(key.createdAt),
            '',
        ));

        apiKeysPage.search = page.search;
        apiKeysPage.limit = page.limit;
        apiKeysPage.order = page.order;
        apiKeysPage.orderDirection = page.orderDirection;
        apiKeysPage.pageCount = page.pageCount;
        apiKeysPage.currentPage = page.currentPage;
        apiKeysPage.totalCount = page.totalCount;

        return apiKeysPage;
    }
}
