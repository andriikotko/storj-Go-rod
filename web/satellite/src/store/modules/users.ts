// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/store';
import { UpdatedUser, User, UsersApi } from '@/types/users';
import { MetaUtils } from '@/utils/meta';

export const USER_ACTIONS = {
    UPDATE: 'updateUser',
    GET: 'getUser',
    CLEAR: 'clearUser',
};

export const USER_MUTATIONS = {
    SET_USER: 'setUser',
    UPDATE_USER: 'updateUser',
    CLEAR: 'clearUser',
};

const {
    GET,
    UPDATE,
} = USER_ACTIONS;

const {
    SET_USER,
    UPDATE_USER,
    CLEAR,
} = USER_MUTATIONS;

/**
 * creates users module with all dependencies
 *
 * @param api - users api
 */
export function makeUsersModule(api: UsersApi): StoreModule<User> {
    return {
        state: new User(),

        mutations: {
            [SET_USER](state: User, user: User): void {
                state.id = user.id;
                state.email = user.email;
                state.shortName = user.shortName;
                state.fullName = user.fullName;
                state.partnerId = user.partnerId;

                if (user.projectLimit === 0) {
                    const limitFromConfig = MetaUtils.getMetaContent('default-project-limit');

                    state.projectLimit = parseInt(limitFromConfig);

                    return;
                }

                state.projectLimit = user.projectLimit;
            },

            [CLEAR](state: User): void {
                state.id = '';
                state.email = '';
                state.shortName = '';
                state.fullName = '';
                state.partnerId = '';
                state.projectLimit = 1;
            },

            [UPDATE_USER](state: User, user: UpdatedUser): void {
                state.fullName = user.fullName;
                state.shortName = user.shortName;
            },
        },

        actions: {
            [UPDATE]: async function ({commit}: any, userInfo: UpdatedUser): Promise<void> {
                await api.update(userInfo);

                commit(UPDATE_USER, userInfo);
            },
            [GET]: async function ({commit}: any): Promise<User> {
                const user = await api.get();

                commit(SET_USER, user);

                return user;
            },
            [CLEAR]: function({commit}: any) {
                commit(CLEAR);
            },
        },

        getters: {
            user: (state: User): User => state,
            userName: (state: User): string => state.getFullName(),
        },
    };
}
