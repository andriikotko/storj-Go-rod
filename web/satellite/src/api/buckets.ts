// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { BaseGql } from '@/api/baseGql';
import { Bucket, BucketCursor, BucketPage, BucketsApi } from '@/types/buckets';

/**
 * BucketsApiGql is a graphql implementation of Buckets API.
 * Exposes all bucket-related functionality.
 */
export class BucketsApiGql extends BaseGql implements BucketsApi {
    /**
     * Fetch buckets.
     *
     * @returns BucketPage
     * @throws Error
     */
    public async get(projectId: string, before: Date, cursor: BucketCursor): Promise<BucketPage> {
        const query =
            `query($projectId: String!, $before: DateTime!, $limit: Int!, $search: String!, $page: Int!) {
                project(id: $projectId) {
                    bucketUsages(before: $before, cursor: {
                        limit: $limit, search: $search, page: $page
                    }) {
                        bucketUsages {
                            bucketName,
                            storage,
                            egress,
                            objectCount,
                            since,
                            before
                        },
                        search,
                        limit,
                        offset,
                        pageCount,
                        currentPage,
                        totalCount
                    }
                }
            }`;

        const variables = {
            projectId,
            before: before.toISOString(),
            limit: cursor.limit,
            search: cursor.search,
            page: cursor.page,
        };

        const response = await this.query(query, variables);

        return this.getBucketPage(response.data.project.bucketUsages);
    }

    /**
     * Method for mapping buckets page from json to BucketPage type.
     *
     * @param page anonymous object from json
     */
    private getBucketPage(page: any): BucketPage {
        if (!page) {
            return new BucketPage();
        }

        const buckets: Bucket[] = page.bucketUsages.map(key =>
            new Bucket(
                key.bucketName,
                key.storage,
                key.egress,
                key.objectCount,
                new Date(key.since),
                new Date(key.before)));

        return new BucketPage(buckets, page.search, page.limit, page.offset, page.pageCount, page.currentPage, page.totalCount);
    }
}
