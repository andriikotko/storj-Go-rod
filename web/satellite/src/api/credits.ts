// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { BaseGql } from '@/api/baseGql';
import { CreditsApi, CreditUsage } from '@/types/credits';

/**
 * CreditsApiGql is a graphql implementation of Credits API.
 * Exposes all credits-related functionality
 */
export class CreditsApiGql extends BaseGql implements CreditsApi {
    /**
     * Fetch CreditUsage
     *
     * @returns CreditUsage
     * @throws Error
     */
    public async get(): Promise<CreditUsage> {
        const query =
            `query {
                creditUsage {
                    referred,
                    usedCredit,
                    availableCredit,
                }
            }`;

        const response = await this.query(query);

        return this.fromJson(response.data.creditUsage);
    }

    /**
     * Method for mapping credit usage from json to CreditUsage type.
     *
     * @param jsonCredits anonymous object from json
     */
    private fromJson(jsonCredits): CreditUsage {
        return new CreditUsage(jsonCredits.referred, jsonCredits.usedCredit, jsonCredits.availableCredit);
    }
}
