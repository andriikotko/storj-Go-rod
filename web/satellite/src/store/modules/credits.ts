// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/store';
import { CreditsApi, CreditUsage } from '@/types/credits';

export const CREDIT_USAGE_ACTIONS = {
    FETCH: 'fetchCreditUsage',
    CLEAR: 'clearCreditUsage',
};

export const CREDIT_USAGE_MUTATIONS = {
    SET: 'SET_CREDIT_USAGE',
    CLEAR: 'CLEAR_CREDIT_USAGE',
};

const { FETCH } = CREDIT_USAGE_ACTIONS;
const { SET, CLEAR } = CREDIT_USAGE_MUTATIONS;

export function makeCreditsModule(api: CreditsApi): StoreModule<CreditUsage> {
    return {
        state: new CreditUsage(),

        mutations: {
            [SET](state: CreditUsage, creditUsage: CreditUsage) {
                state.availableCredits = creditUsage.availableCredits;
                state.usedCredits = creditUsage.usedCredits;
                state.referred = creditUsage.referred;
            },
            [CLEAR](state: CreditUsage): void {
                state.availableCredits = 0;
                state.usedCredits = 0;
                state.referred = 0;
            },
        },

        actions: {
            [FETCH]: async function({commit}: any): Promise<CreditUsage> {
                const credits = await api.get();

                commit(SET, credits);

                return credits;
            },
            [CREDIT_USAGE_ACTIONS.CLEAR]: function({commit}: any): void {
                commit(CLEAR);
            },
        },

        getters: {
            credits: (state: CreditUsage): CreditUsage => state,
        },
    };
}
