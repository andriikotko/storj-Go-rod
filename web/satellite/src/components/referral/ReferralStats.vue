// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="referral-stats">
        <p class="referral-stats__title"> {{ title }} </p>
        <div class="referral-stats__wrapper">
            <div class="referral-stats__card"
                v-for="(stat, key) in stats"
                :key="key"
                :style="stat.background">
                <span class="referral-stats__card-text">
                    <span class="referral-stats__card-title">{{ stat.title }}</span>
                    <span class="referral-stats__card-description">{{ stat.description }}</span>
                </span>
                <br>
                <span class="referral-stats__card-number">{{ stat.symbol + usage[key] }}</span>
            </div>
        </div>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import { CREDIT_USAGE_ACTIONS } from '@/store/modules/credits';
import { CreditUsage } from '@/types/credits';

class CreditDescription {
    public title: string;
    public description: string;
    public symbol: string;

    // possibly we could add some 'style' type
    public style: any;

    constructor(title: string, description: string, symbol: string, color: string) {
        this.title = title;
        this.description = description;
        this.symbol = symbol;
        this.style = { backgroundColor: color };
    }
}

@Component
export default class ReferralStats extends Vue {
    private readonly TITLE_SUFFIX: string = 'Here Are Your Referrals So Far';
    private stats: CreditDescription[] = [
        new CreditDescription('referrals made', 'People you referred who signed up', '', '#FFFFFF'),
        new CreditDescription('earned credits', 'Free credits that will apply to your upcoming bill', '$', 'rgba(217, 225, 236, 0.5)'),
        new CreditDescription('applied credits', 'Free credits that have already been applied to your bill', '$', '#D1D7E0'),
    ];

    public async mounted() {
        // TODO: we pre-fetch all data in /src/views/Dashboard, but this is tardigrade related, so could be here
        await this.fetch();
    }

    public async fetch(): Promise<void> {
        try {
            await this.$store.dispatch(CREDIT_USAGE_ACTIONS.FETCH);
        } catch (error) {
            await this.$notify.error('Unable to fetch credit usage: ' + error.message);
        }
    }

    public get title(): string {
        // TODO: not sure that we are able to create a user with empty name
        const name = this.$store.state.usersModule.fullName || '' ;

        return name.length > 0 ? `${name}, ${this.TITLE_SUFFIX}` : this.TITLE_SUFFIX;
    }

    public get usage(): CreditUsage {
        return this.$store.getters.credits;
    }
}
</script>

<style scoped lang="scss">
    .referral-stats {

        &__title {
            text-align: center;
            font-family: 'font_bold', sans-serif;
        }

        &__wrapper {
            display: flex;
            flex-direction: row;
            justify-content: space-around;
            left: 15%;
            right: 15%;
            font-family: 'font_regular', sans-serif;
        }

        &__card {
            color: #354049;
            min-height: 176px;
            max-width: 276px;
            flex-basis: 25%;
            border-radius: 24px;
            padding-top: 25px;
            padding-left: 26px;
            padding-right: 29px;
            display: flex;
            flex-direction: column;
            justify-content: space-evenly;

            &-text {
                display: block;
            }

            &-title {
                display: block;
                text-transform: uppercase;
                font-family: 'font_bold', sans-serif;
                font-size: 14px;
                line-height: 18px;
            }

            &-description {
                display: block;
                font-size: 14px;
                line-height: 21px;
                margin-top: 7px;
            }

            &-number {
                display: block;
                font-family: 'font_bold', sans-serif;
                font-size: 46px;
                line-height: 60px;
                margin-bottom: 27px;
            }
        }
    }
</style>
