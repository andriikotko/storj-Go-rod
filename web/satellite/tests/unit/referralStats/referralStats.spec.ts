// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';

import ReferralStats from '@/components/referral/ReferralStats.vue';

import { makeCreditsModule } from '@/store/modules/credits';
import { makeUsersModule, USER_ACTIONS } from '@/store/modules/users';
import { CreditUsage } from '@/types/credits';
import { User } from '@/types/users';
import { createLocalVue, mount } from '@vue/test-utils';

import { CreditsApiMock } from '../mock/api/credits';
import { UsersApiMock } from '../mock/api/users';

const {
    GET,
} = USER_ACTIONS;

const localVue = createLocalVue();
localVue.use(Vuex);

const mockUser = new User('1', 'full name', 'short name');
const mockCredits = new CreditUsage(1, 2, 3);

const creditsApi = new CreditsApiMock();
creditsApi.setMockCredits(mockCredits);
const usersApi = new UsersApiMock();
usersApi.setMockUser(mockUser);

const usersModule = makeUsersModule(usersApi);
const creditsModule = makeCreditsModule(creditsApi);

const store = new Vuex.Store({ modules: { usersModule, creditsModule } });

store.dispatch(GET);

describe('ReferralStats', () => {
    it('fetch method populate store', async () => {
        const wrapper = mount(ReferralStats, {
            store,
            localVue,
        });

        await wrapper.vm.fetch();

        const storeState = (store.state as any).creditsModule;

        expect(storeState.referred).toBe(mockCredits.referred);
        expect(storeState.usedCredits).toBe(mockCredits.usedCredits);
        expect(storeState.availableCredits).toBe(mockCredits.availableCredits);
    });

    it('snapshot not changed', () => {
        const wrapper = mount(ReferralStats, {
            store,
            localVue,
        });

        expect(wrapper).toMatchSnapshot();
    });

    it('title renders correctly', () => {
        const wrapper = mount(ReferralStats, {
            store,
            localVue,
        });

        const title = wrapper.find('.referral-stats__title');

        expect(title.text()).toBe(`${mockUser.fullName}, ${wrapper.vm.$data.TITLE_SUFFIX}`);
    });

    it('credits type title renders correctly', () => {
        const wrapper = mount(ReferralStats, {
            store,
            localVue,
        });

        const titles = wrapper.findAll('.referral-stats__card-title');

        expect(titles.at(0).text()).toBe(wrapper.vm.$data.stats[0].title);
        expect(titles.at(1).text()).toBe(wrapper.vm.$data.stats[1].title);
        expect(titles.at(2).text()).toBe(wrapper.vm.$data.stats[2].title);
    });

    it('credits type description renders correctly', () => {
        const wrapper = mount(ReferralStats, {
            store,
            localVue,
        });

        const descriptions = wrapper.findAll('.referral-stats__card-description');

        expect(descriptions.at(0).text()).toBe(wrapper.vm.$data.stats[0].description);
        expect(descriptions.at(1).text()).toBe(wrapper.vm.$data.stats[1].description);
        expect(descriptions.at(2).text()).toBe(wrapper.vm.$data.stats[2].description);
    });

    it('credits type symbol with value renders correctly', () => {
        const wrapper = mount(ReferralStats, {
            store,
            localVue,
        });

        const descriptions = wrapper.findAll('.referral-stats__card-number');

        // TODO: write assertions when 'stat.symbol + usage[key]' will be fixed
    });
});
