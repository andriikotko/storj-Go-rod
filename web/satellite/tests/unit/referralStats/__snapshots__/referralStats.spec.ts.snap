// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReferralStats snapshot not changed 1`] = `
<div class="referral-stats">
  <p class="referral-stats__title"> full name, Here Are Your Referrals So Far </p>
  <div class="referral-stats__wrapper">
    <div class="referral-stats__card"><span class="referral-stats__card-text"><span class="referral-stats__card-title">referrals made</span> <span class="referral-stats__card-description">People you referred who signed up</span></span> <br> <span class="referral-stats__card-number">undefined</span></div>
    <div class="referral-stats__card"><span class="referral-stats__card-text"><span class="referral-stats__card-title">earned credits</span> <span class="referral-stats__card-description">Free credits that will apply to your upcoming bill</span></span> <br> <span class="referral-stats__card-number">$undefined</span></div>
    <div class="referral-stats__card"><span class="referral-stats__card-text"><span class="referral-stats__card-title">applied credits</span> <span class="referral-stats__card-description">Free credits that have already been applied to your bill</span></span> <br> <span class="referral-stats__card-number">$undefined</span></div>
  </div>
</div>
`;
